torch/nn/_functions/thnn/sparse.py

```diff
@@ -103,3 +103,155 @@ def backward(self, grad_output):
 
 
 _all_functions.append(Embedding)
+
+MODE_SUM = 0
+MODE_MEAN = 1
+
+
+class EmbeddingBag(Function):
+
+    def __init__(self, max_norm, norm_type, scale_grad_by_freq, mode):
+        super(EmbeddingBag, self).__init__()
+        self.max_norm = max_norm
+        self.norm_type = norm_type
+        self.scale_grad_by_freq = scale_grad_by_freq
+        self._indices = None
+        assert mode is not None
+        if mode == 'sum':
+            self.mode = MODE_SUM
+        elif mode == 'mean':
+            self.mode = MODE_MEAN
+        else:
+            raise ValueError("mode needs to be 'sum' or 'mean', but got {}"
+                             .format(mode))
+
+    def _renorm(self, indices, weight):
+        self._backend.LookupTable_renorm(
+            self._backend.library_state,
+            indices,
+            weight,
+            self.max_norm,
+            self.norm_type
+        )
+
+    def forward(self, weight, indices, offsets):
+        assert not self.needs_input_grad[1], "EmbeddingBag doesn't " \
+            "compute the gradient w.r.t. the indices"
+
+        assert not self.needs_input_grad[2], "EmbeddingBag doesn't " \
+            "compute the gradient w.r.t. the offsets"
+
+        assert indices.dim() == 1
+        if offsets.dim() != 1:
+            raise ValueError("offsets has to be a 1D Tensor")
+
+        if offsets[0] != 0:
+            raise ValueError("offsets[0] has to be 0, i.e. the first sequence"
+                             " in the mini-batch has to start from position 0."
+                             "However, got {}".format(offsets[0]))
+        if offsets[-1] > indices.size(0):
+            raise ValueError("offsets[-1] has to be smaller than indices's length"
+                             " ({}), but got offsets[-1] of {}"
+                             .format(indices.size(0), offsets[-1]))
+
+        self._backend = type2backend[type(weight)]
+        self._weight_size = weight.size()
+        self._offset2bag = offsets.new()
+
+        self.save_for_backward(indices)
+
+        indices = indices.contiguous().view(-1)
+        output = weight.new()
+        if self.max_norm is not None:
+            self._renorm(indices, weight)
+
+        if weight.is_cuda:
+            if self.mode == MODE_MEAN:
+                self.sequence_length = offsets.new().resize_(offsets.size())
+            else:
+                self.sequence_length = None
```

Comment:
okay changing. I also found via contbuild that the corner-case of 1 sequence (i.e. offsets.size(0) == 1) is not handled on the CPU side. I'll fix that as well.